### PR TITLE
xeth, core, event/filter, rpc: new block and transaction filters

### PR DESCRIPTION
**Pull request by obscuren**

### CODE REVIEW

**Comment by frozeman** [comment]
`eth_newBlockFilter` works, but for `eth_newPendingTransactionFilter` seems not to be implemented

File: xeth/xeth.go [hunk 1]
```diff
@@ -370,71 +429,90 @@ func (self *XEth) RegisterFilter(earliest, latest int64, skip, max int, address
...
 	return id
 }
 
-func (self *XEth) UninstallFilter(id int) bool {
-	if _, ok := self.logs[id]; ok {
-		delete(self.logs, id)
-		self.filterManager.UninstallFilter(id)
-		return true
-	}
+func (self *XEth) NewTransactionFilter() int {
```
Comment (frozeman) [inline xeth/xeth.go]: This should probably be `NewPendingTransactionFilter `

**Comment by frozeman** [comment]
I currently can't run the RPC test as the blocktest is not loading.

File: xeth/xeth.go (hunk 1 above)
Comment (obscuren) [inline xeth/xeth.go]: That's not the line you're looking for. But indeed, I implemented it as `eth_newTransactionFilter`. I'll change it to `eth_newPendingTransactionFilter`

Comment (frozeman) [inline xeth/xeth.go]: I tried both. Are sure that its also exposed to the RPC?

**Comment by obscuren** [comment]
I think this is now properly implemented:

Create filter

```
curl --data '{"jsonrpc":"2.0","method":"eth_newPendingTransactionFilter","params":[],"id":1}' http://127.0.0.1:8545
{
	"id": 5,
	"jsonrpc": "2.0",
	"result": "0x0"
}
```

Send transaction

```
> eth.sendTransaction({from: eth.accounts[0], to: eth.accounts[1], value: "10", gasPrice: "10"})
'0x3e477f3c0b66864a2d3775e8889f0f75a35dcb8f3d40c63a309a77375bccde56'
```

Get filter changes

```
curl --data '{"jsonrpc":"2.0","method":"eth_getFilterChanges","params":["0x0"],"id":5}' http://127.0.0.1:8545
{
	"id": 5,
	"jsonrpc": "2.0",
	"result": [
		"0x3e477f3c0b66864a2d3775e8889f0f75a35dcb8f3d40c63a309a77375bccde56"
	]
}
```

Get the transactions (including the pending)

```
> eth.getTransaction("0x3e477f3c0b66864a2d3775e8889f0f75a35dcb8f3d40c63a309a77375bccde56")
{
  transactionIndex: 0,
  value: '10',
  input: '0x',
  nonce: 329,
  blockHash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  from: '0x1da8b4c338fcf6d64e64383f5a8d4da86eea9a6d',
  to: '0x30867b568b4cf518d4c432e8b140e1010e9f021c',
  gas: 90000,
  gasPrice: '10',
  hash: '0x3e477f3c0b66864a2d3775e8889f0f75a35dcb8f3d40c63a309a77375bccde56',
  blockNumber: 0
}
```

**Comment by frozeman** [comment]
will test in an h, thanks!!

**Comment by frozeman** [comment]
All seems to work fine!

But i have two things:
- When i use `eth_getTransactionByHash` the console shows the following error `I0508 09:42:10.729635   84136 xeth.go:257] EOF`
- I would strongly advise that we return `null` for the `blockNumber` and `blockHash` in the pending transaction. We agreed to that when we been here all together and i think its very necessary to allow for reliable checks.
if you get a transaction using `web3.eth.getTransaction('0x1235...')` its hard to tell if its pending or not. Except you compare the `blockNumber === 0`, which imho is kind of wrong, as block number 0 is the genisis block. (This means we would have to explain that in the docs, that 0 means pending, whereas `null` is self explanatory)

I would do the same with pending logs (which rely on https://github.com/ethereum/go-ethereum/issues/852 to be fixed).

**Comment by obscuren** [comment]
No. `null` is not a valid hash neither a number.

**Comment by obscuren** [comment]
This PR will be merged because the change has been addressed. We can open a new issue for any pending (other) issues.

**Pull request merged**